### ISSUE TITLE
Poor integrity check performance at high cardinalities

### ISSUE THREAD

**Issue by DanielThomas** [Netflix/hollow#312]
We have a list type with `63,457,461` records with a heap footprint of 1.1 GB. Producer checksums for these types add over 2 minutes at the end of the producer run, because the individual checksums for these takes ~35 seconds, times the four integrity checks:

```
1540230505469 - start checksum for - ListOfModuleDependency List<ModuleDependency>;
1540230539884 - end checksum for - ListOfModuleDependency List<ModuleDependency>;
```

Looks like all `Hollow*TypeReadState` have an inner loop that iterates the populated ordinals, skipping those that aren't in the current shard:

https://github.com/Netflix/hollow/blob/0700bcad817a91d44b317f3f2ad6549a458aaf73/hollow/src/main/java/com/netflix/hollow/core/read/engine/list/HollowListTypeReadStateShard.java#L102-L112

The shards are iterated in the outer loop:

https://github.com/Netflix/hollow/blob/f9873ee423f3c314bcdb79d8cf5513dc4e2602d6/hollow/src/main/java/com/netflix/hollow/core/read/engine/list/HollowListTypeReadState.java#L203-L204

In this case there are 128 shards and it takes ~1.5 seconds per shard:
```
0 - 1540232787697
1 - 1540232789155
2 - 1540232790620
3 - 1540232792088
4 - 1540232793552
5 - 1540232795017
6 - 1540232796488
7 - 1540232797957
8 - 1540232799429
9 - 1540232800898
10 - 1540232802376
...
```

Appears that `BitSet.nextSetBit` is not an insignificant contributor, so calculating the shard ordinals once in the outer call and passing them in seems to be a quick win - this is a sampling profiler however, and with such short method call durations microbenchmarks are probably a better way of getting the contribution of those inner calls:

<img width="2026" alt="screen shot 2018-10-22 at 1 05 06 pm" src="https://user-images.githubusercontent.com/1479220/47316086-23dd1880-d5fb-11e8-81e0-b5700ce996f7.png">

Issue https://github.com/Netflix/hollow/issues/306 would also avoid this for us, some more details about the size of the data there.

**Comment by DanielThomas** [comment]
Here's some timings from our production producer - 40% of the producer time is spent in integrity checks:

```
11:00:49.340 starting cycle
11:01:05.158 populating data...
11:02:31.875 data populated; duration=86s
11:02:31.876 publishing blobs...
11:03:02.811 delta published; fromVersion=20190129185545018 toVersion=20190129190049019 source=/mnt/extra-storage/astrid/delta-20190129185545018-20190129190049019.7e85da78 size=6016 duration=0s
11:03:03.007 reversedelta published; fromVersion=20190129190049019 toVersion=20190129185545018 source=/mnt/extra-storage/astrid/reversedelta-20190129190049019-20190129185545018.1727af4a size=4052 duration=0s
11:03:46.301 snapshot published; fromVersion=none toVersion=20190129190049019 source=/mnt/extra-storage/astrid/snapshot-20190129190049019.4698bb14 size=1360135256 duration=43s
11:03:46.302 delta, reversedelta & snapshot blobs published; duration=74s
11:03:46.303 checking blob integrity...
11:05:50.766 blob integrity good; duration=124s
11:05:50.767 running validations...
11:05:50.768 no validations run; duration=0s
11:05:50.768 announcing...
11:05:50.797 announced; duration=0s
11:05:50.917 cycle succeeded; duration=301s
```

**Comment by PaulSandoz** [comment]
- It should be possible to create just one loop iterating over the populated ordinals, selecting the shard and calling a new method for the checksum with a given ordinal. The side-effect of that approach might be a lack of cache coherency when operating on the shard read state.  Although it's not obvious the other way is any better regarding coherency given how the ordinals are not bucketed per shard but interleaved.
- It should also be possible to skip through the bit set faster with `nextSetBit` from a known matching shard ordinal with `ordinal + numShards`, or a non-matching ordinal rounded up to the nearest potential matching shard ordinal. 

In either case need microbenchmarks as that might show up more optimization opportunities.

**Comment by PaulSandoz** [comment]
Fixed in #397
